pandas/core/reshape.py

```diff
@@ -447,46 +465,35 @@ def get_compressed_ids(labels, sizes):
     return _compress_group_index(ids, sort=True)
 
 
-def stack(frame, level=-1, dropna=True):
+def stack_single_level(frame, level_num, dropna=True):
     """
-    Convert DataFrame to Series with multi-level Index. Columns become the
-    second level of the resulting hierarchical index
+    Convert DataFrame to DataFrame or Series with multi-level Index.
+    Columns become the second level of the resulting hierarchical index
 
     Returns
     -------
-    stacked : Series
+    stacked : DataFrame or Series
     """
-    N, K = frame.shape
     if isinstance(frame.columns, MultiIndex):
-        if frame.columns._reference_duplicate_name(level):
-            msg = ("Ambiguous reference to {0}. The column "
-                   "names are not unique.".format(level))
-            raise ValueError(msg)
+        return stack_multi_levels_simultaneously(frame, level_nums=[level_num], dropna=dropna)
 
-    # Will also convert negative level numbers and check if out of bounds.
-    level_num = frame.columns._get_level_number(level)
-
-    if isinstance(frame.columns, MultiIndex):
-        return _stack_multi_columns(frame, level_num=level_num, dropna=dropna)
-    elif isinstance(frame.index, MultiIndex):
+    # frame.columns is a simple Index (not a MultiIndex)
+    N, K = frame.shape
+    if isinstance(frame.index, MultiIndex):
         new_levels = list(frame.index.levels)
-        new_levels.append(frame.columns)
-
         new_labels = [lab.repeat(K) for lab in frame.index.labels]
-        new_labels.append(np.tile(np.arange(K), N).ravel())
-
         new_names = list(frame.index.names)
-        new_names.append(frame.columns.name)
-        new_index = MultiIndex(levels=new_levels, labels=new_labels,
-                               names=new_names, verify_integrity=False)
     else:
-        ilabels = np.arange(N).repeat(K)
-        clabels = np.tile(np.arange(K), N).ravel()
-        new_index = MultiIndex(levels=[frame.index, frame.columns],
-                               labels=[ilabels, clabels],
-                               names=[frame.index.name, frame.columns.name],
-                               verify_integrity=False)
-
+        idx_labels, new_levels = factorize(frame.index)
```

Comment:
I guess one could simply implement ``to_hierarchical`` on the base ``Index``?

In this particular usage, since I need the ``levels``, ``labels``, and ``names`` separately, and would be discarding immediately the ``mi = index.to_heirarchical(N)``, I'm not sure the slight simplification of the code (which is just a few contiguous lines) justifies the performance hit of creating a new ``MultiIndex`` just to discard it immediately, but if you think it's worth it, I'll give it a shot.